app/Jobs/MarkNotificationsRead.php

```diff
@@ -31,56 +28,44 @@ public function __construct($object, $user)
 
-    public function forForumPost()
-    {
-        $this->notifiable = $this->object->topic()->withTrashed()->first();
-
-        if ($this->notifiable === null) {
-            throw new Exception("Can't find topic {$this->object->getKey()} of post {$this->object->getKey()}");
-        }
-
-        $this->notificationTime = $this->object->post_time;
-    }
-
     public function handle()
     {
-        try {
-            if ($this->object instanceof Beatmapset) {
-                // do nothing
-            } elseif ($this->object instanceof ForumPost) {
-                $this->forForumPost();
-            } else {
-                throw new Exception('Unknown object to be marked as read: '.get_class($this->object));
-            }
-        } catch (Exception $e) {
-            log_error($e);
-
-            return;
+        if (!($this->object instanceof ForumPost)) {
+            throw new Exception('Unknown object to be marked as read: '.get_class($this->object));
         }
 
-        if (!isset($this->notifiable)) {
-            $this->notifiable = $this->object;
-        }
+        $notifiable = $this->object->topic()->withTrashed()->first();
 
-        if (!isset($this->notificationTime)) {
-            $this->notificationTime = now();
+        if ($notifiable === null) {
+            throw new Exception("Can't find topic {$this->object->getKey()} of post {$this->object->getKey()}");
         }
 
+        $notifiableId = $notifiable->getKey();
+        $notifiableType = MorphMap::getType($notifiable);
+
         $notifications = Notification
-            ::where('notifiable_type', '=', MorphMap::getType($this->notifiable))
-            ->where('notifiable_id', '=', $this->notifiable->getKey())
-            ->where('created_at', '<=', $this->notificationTime);
+            ::where('notifiable_type', '=', $notifiableType)
+            ->where('notifiable_id', '=', $notifiableId)
+            ->where('created_at', '<=', $this->object->post_time);
 
         $userNotifications = $this->user
             ->userNotifications()
             ->where('is_read', '=', false)
             ->whereIn('notification_id', $notifications->select('id'));
 
-        $notificationIds = $userNotifications->pluck('notification_id')->all();
-        $userNotifications->update(['is_read' => true]);
-
-        if (!empty($notificationIds)) {
-            event(new NotificationReadEvent($this->user->getKey(), ['ids' => $notificationIds]));
+        $count = $userNotifications->update(['is_read' => true]);
+        $notificationIdentity = [
+            'category' => Notification::nameToCategory(Notification::FORUM_TOPIC_REPLY),
+            'object_id' => $notifiableId,
+            'object_type' => $notifiableType,
+        ];
+
+        if ($count > 0) {
+            event(new NotificationReadEvent($this->user->getKey(), [
+                'notifications' => [$notificationIdentity],
+                'read_count' => $count,
+                'timestamp' => now(),
```

Comment:
it looks like this will mark everything client-side as read even though it's only up to the post's `post_time` server-side? (which may be quite a bit before `now()` and the topic may have further unread posts)